scapy/contrib/rtps/common_types.py

```diff
@@ -0,0 +1,323 @@
+        return end
+
+    return False
+
+
+def e_flags(pkt: Packet) -> str:
+    if is_le(pkt):
+        return FORMAT_LE
+    else:
+        return FORMAT_BE
+
+
+class EField(object):
+    """
+    A field that manages endianness of a nested field passed to the constructor
+    """
+
+    __slots__ = ["fld", "endianness", "endianness_from"]
+
+    def __init__(self, fld, endianness=FORMAT_BE, endianness_from=e_flags):
+        self.fld = fld
+        self.endianness = endianness
+        self.endianness_from = endianness_from
+
+    def set_endianness(self, pkt):
+        if getattr(pkt, "endianness", None) is not None:
+            self.endianness = pkt.endianness
+        elif self.endianness_from is not None:
+            self.endianness = self.endianness_from(pkt)
+
+        if hasattr(self.fld, "set_endianness"):
+            self.fld.set_endianness(endianness=self.endianness)
+            return
+
+        if hasattr(self.fld, "endianness"):
+            self.fld.endianness = self.endianness
+            return
+
+        if isinstance(self.endianness, str) and self.endianness:
+            if hasattr(self.fld, "fmt"):
+                if len(self.fld.fmt) == 1:  # if it's only "I"
+                    _end = self.fld.fmt[0]
+                else:  # if it's "<I"
+                    _end = self.fld.fmt[1:]
+                self.fld.fmt = self.endianness + _end
+
+                self.fld.struct = struct.Struct(self.fld.fmt)
+
+    def getfield(self, pkt, buf):
+        self.set_endianness(pkt)
+        return self.fld.getfield(pkt, buf)
+
+    def addfield(self, pkt, buf, val):
+        self.set_endianness(pkt)
+        return self.fld.addfield(pkt, buf, val)
+
+    def randval(self):
+        return self.fld.randval()
+
+    def __getattr__(self, attr):
+        return getattr(self.fld, attr)
+
+
+class EPacket(Packet):
+    """A packet that manages its endianness"""
+
+    __slots__ = ["endianness"]
+
+    def __init__(self, *args, endianness=None, **kwargs):
+        self.endianness = endianness
+        super().__init__(*args, **kwargs)
+
+    def extract_padding(self, p):
+        return b"", p
+
+
+class EPacketField(PacketField):
+    """
+    A packet field that manages its endianness and that of its nested packet
+    """
+
+    __slots__ = ["endianness", "endianness_from", "fuzz_fun"]
+
+    def __init__(
+        self,
+        *args,
+        fuzz_fun=fuzz,
+        endianness=None,
+        endianness_from=e_flags,
+        **kwargs,
+    ):
+        self.endianness = endianness
+        self.endianness_from = endianness_from
+        self.fuzz_fun = fuzz_fun
+        super().__init__(*args, **kwargs)
+
+    def set_endianness(self, pkt):
+        if getattr(pkt, "endianness", None) is not None:
+            self.endianness = pkt.endianness
+        elif self.endianness_from is not None and pkt:
+            self.endianness = self.endianness_from(pkt)
+            assert self.endianness is not None
```

Comment:
It think that it is better to replace this assert by a warning and a default value.

Reply:
I'm about to commit this:

```python
    def set_endianness(self, pkt):
        if getattr(pkt, "endianness", None) is not None:
            self.endianness = pkt.endianness
        elif self.endianness_from is not None and pkt:
            self.endianness = self.endianness_from(pkt)
            if self.endianness is None:
                warnings.warn(
                    'Endianess should never be None.'
                    'Setting it to default: {}', DEFAULT_ENDIANESS)
                self.endianness = DEFAULT_ENDIANESS
```